### DISCUSSION TITLE
UserData for Windows VMs?

### DISCUSSION THREAD

**Discussion by btzq** [apache/cloudstack#9584] [Q&A] [answered]
Hi Guys, 

Is there anyway to use UserData for Windows VMs?

We have to Autoscale a bunch of Windows Guest VM but its painful to do so without having to use UserData. 

We dont have issues with Linux VMs

**Comment by weizhouapache** [comment]
cloudbase-init may help you
https://github.com/cloudbase/cloudbase-init

**Comment by btzq** [comment]
We tried this but it didnt work for us.

- We installed Cloudbase in the Windows Guest, and turned it off
- Added User Data script below
- Turn on Windows Guest VM
- Checked inside VM, and saw that the userdata did not take effect. 

We tried 4 different userdata scripts and all did not work:

![Screenshot 2024-08-24 at 8 17 25 AM](https://github.com/user-attachments/assets/f7d943a1-58db-4523-a076-830dac58b296)

Did we missing a step?

**Reply by weizhouapache** [reply]
> We tried this but it didnt work for us.
> 
> - We installed Cloudbase in the Windows Guest, and turned it off
> - Added User Data script below
> - Turn on Windows Guest VM
> - Checked inside VM, and saw that the userdata did not take effect. 
> 
> We tried 4 different userdata scripts and all did not work:
> 
> ![Screenshot 2024-08-24 at 8 17 25 AM](https://github.com/user-attachments/assets/f7d943a1-58db-4523-a076-830dac58b296)
> 
> Did we missing a step?

You may need to change its configuration to support cloudstack, similar as DataSources of cloud-init

**Reply by Ye-Min-Tun** [reply]
> > We tried this but it didnt work for us.
> > 
> > * We installed Cloudbase in the Windows Guest, and turned it off
> > * Added User Data script below
> > * Turn on Windows Guest VM
> > * Checked inside VM, and saw that the userdata did not take effect.
> > 
> > We tried 4 different userdata scripts and all did not work:
> > ![Screenshot 2024-08-24 at 8 17 25 AM](https://private-user-images.githubusercontent.com/147712084/361088302-f7d943a1-58db-4523-a076-830dac58b296.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUyODI3MzMsIm5iZiI6MTcyNTI4MjQzMywicGF0aCI6Ii8xNDc3MTIwODQvMzYxMDg4MzAyLWY3ZDk0M2ExLTU4ZGItNDUyMy1hMDc2LTgzMGRhYzU4YjI5Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwMlQxMzA3MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMzc1NTc0ZDVkZTBkNzAzYTEzYzc1MmJkYjM2MGJlYjg2MTYxNDE2YzkwZTA4MjUzMGQzZmFiMDIzYzM2OTM2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9._WDzKbAvlEkoNu9s6iUozTVSN5FxapmGixbMjzBQ73c)
> > Did we missing a step?
> 
> You may need to change its configuration to support cloudstack, similar as DataSources of cloud-init

@weizhouapache Hi, I'm also looking for userdata support for Windows autoscaled VMs in cloudstack. 
Can you pls explain more details about what need to change the configuration? Do you mean cloudbase-init config?

**Reply by weizhouapache** [reply]
yes @Ye-Min-Tun 
You can refer to cloudbase-init doc
https://cloudbase-init.readthedocs.io/en/latest/services.html

` metadata_services` needs to be set to `cloudbaseinit.metadata.services.cloudstack.CloudStack` to support cloudstack metadata and password server.

**Reply by Ye-Min-Tun** [reply]
@weizhouapache I set the `metadata_services`, but still didn't work. What did I miss to configure? Here's the `cloudbase-init-unattend.conf `
```
[DEFAULT]
username=Admin
groups=Administrators
inject_user_password=true
config_drive_raw_hhd=true
config_drive_cdrom=true
config_drive_vfat=true
bsdtar_path=C:\Program Files\Cloudbase Solutions\Cloudbase-Init\bin\bsdtar.exe
mtools_path=C:\Program Files\Cloudbase Solutions\Cloudbase-Init\bin\
verbose=true
debug=true
log_dir=C:\Program Files\Cloudbase Solutions\Cloudbase-Init\log\
log_file=cloudbase-init-unattend.log
default_log_levels=comtypes=INFO,suds=INFO,iso8601=WARN,requests=WARN
logging_serial_port_settings=COM1,115200,N,8
mtu_use_dhcp_config=true
ntp_use_dhcp_config=true
local_scripts_path=C:\Program Files\Cloudbase Solutions\Cloudbase-Init\LocalScripts\
check_latest_version=false
metadata_services=cloudbaseinit.metadata.services.cloudstack.CloudStack
plugins = cloudbaseinit.plugins.common.mtu.MTUPlugin, cloudbaseinit.plugins.windows.ntpclient.NTPClientPlugin, cloudbaseinit.plugins.common.sethostname.SetHostNamePlugin, cloudbaseinit.plugins.windows.createuser.CreateUserPlugin, cloudbaseinit.plugins.common.networkconfig.NetworkConfigPlugin, cloudbaseinit.plugins.windows.licensing.WindowsLicensingPlugin, cloudbaseinit.plugins.common.sshpublickeys.SetUserSSHPublicKeysPlugin, cloudbaseinit.plugins.windows.extendvolumes.ExtendVolumesPlugin, cloudbaseinit.plugins.common.userdata.UserDataPlugin, cloudbaseinit.plugins.common.setuserpassword.SetUserPasswordPlugin, cloudbaseinit.plugins.windows.winrmlistener.ConfigWinRMListenerPlugin, cloudbaseinit.plugins.windows.winrmcertificateauth.ConfigWinRMCertificateAuthPlugin, cloudbaseinit.plugins.common.localscripts.LocalScriptsPlugin
allow_reboot=false
stop_service_on_exit=false

``` 
And I tried this two userdata in Cloudstack.
```
#cloud-config
runcmd:
  - 'powershell.exe -Command "New-Item -Path C:\TestDir -ItemType Directory"'
  - 'powershell.exe -Command "New-Item -Path C:\TestDir\testfile.txt -ItemType File"'
  - 'powershell.exe -Command "Add-Content -Path C:\TestDir\testfile.txt -Value ''Hello, CloudStack!''"'
``` 
```
  
#ps1_sysnative
Set-ExecutionPolicy -ExecutionPolicy Unrestricted -Scope Process -Force
$logPath = "C:\Users\Public\userdata.log"
"Script executed" | Out-File $logPath
``` 
Or is there anything wrong with my userdata script?

**Reply by weizhouapache** [reply, marked as answer]
@Ye-Min-Tun 
I can not help more on this. 
you may ask the cloudbase-init project: https://github.com/cloudbase/cloudbase-init
